powerman: when status and status_all are defined, use status_all only on full pluglist

chu11 opened the pull request:
Problem: When selecting a script to use for an action, the "all" version of a script is always used if the action is a query action (e.g. power status query, beacon query, etc.).

This can be problematic in situations where blades/nodes/etc. are not fully populated, such as in a chassis.  We do not want to query "all" things in that case.

Solution: Only call the "all" script if it is a query and a singlet version of the script does not exist.  For device files where unpopulated nodes may exist, it can define an all script and singlet script for each scenario.

Built on top of #155

garlick commented:
Nice!  This probably needs test coverage though.

chu11 commented:
> Nice! This probably needs test coverage though.

Ahh, yeah.  I guess it was sort of there in the bigger PR, but could use some specific tests targeted just to this.

chu11 commented:
re-pushed, added some coverage in t0004-status-query.t.

garlick commented:
Thanks for the test!

Other things:
- Title should be more descriptive like, "powerman: when status and status_all are defined, use status_all only on full pluglist"
- Any existing dev specs that define both should be updated to only define `status_all` to avoid using formerly dead and probably untested code.  See `appro-gb2.dev`, `appro-greenblade.dev` and `swpdu.dev`, all of which contain multiple specifications.
- This  behavior should be explained in powerman.dev(5).

BTW which script does powerman choose when the device spec contains both `status` and `status_all` and does not define a `plug name` list?

garlick reviewed (commented):
Oh just realized that this applies to `status_beacon_all`.
Hmm, looks like there is a `status_temp_all` but no `status_temp`?

chu11 commented:
> BTW which script does powerman choose when the device spec contains both status and status_all and does not define a plug name list?

TBH I wasn't sure.  Adding some printf debug into  `_enqueue_targeted_actions()` it appears `dev->plugs` is simply all the plugs defined by `node` configuration if the plug name list isn't formally listed.  And if the user doesn't input anything, the input hostlist `hl` parameter is everything specified in the config file.

chu11 commented:
re-pushed ... tweaked nit things per comments above and ...

- added beacon and temperature equivalent script tests, it ends up "status_temp" was never covered in `vpcd` and there was a corner case in there :P
- add documentation in `powerman.dev(5)`
- removed unusued/untested singleton scripts in device files
-

garlick commented on etc/devices/appro-gb2.dev:
```diff
@@ -44,18 +44,6 @@ specification "sr5110" {
 		}
 		expect "-iSCB> "
 	}
-	script status {
-		send "pmnode %s\r"
-		expect "node([0-9]+): (on|off|n/a)"
```
In the commit message for this, it would probably be good to mention that the current behavior is changing so that the untested singletons would start to get used, so this change just preserves the existing behavior.

garlick reviewed (approved):
LGTM! Just one commit message suggestion.

chu11 commented:
thanks, pushed a commit message tweak and setting MWP
